components/brave_federated_learning/brave_operational_profiling.cc

```diff
@@ -0,0 +1,205 @@
+  ephemeral_id_expiration_time_ =
+      pref_service_->GetTime(kEphemeralIdExpirationPrefName);
+}
+
+void BraveOperationalProfiling::SavePrefs() {
+  pref_service_->SetInteger(kLastCheckedSlotPrefName, last_checked_slot_);
+  pref_service_->SetString(kEphemeralIdPrefName, ephemeral_id_);
+  pref_service_->SetTime(kEphemeralIdExpirationPrefName,
+                         ephemeral_id_expiration_time_);
+}
+
+void BraveOperationalProfiling::OnCollectionSlotStartTimerFired() {
+  simulate_local_training_step_timer_->Reset();
+}
+
+void BraveOperationalProfiling::OnSimulateLocalTrainingStepTimerFired() {
+  SendCollectionSlot();
+}
+
+void BraveOperationalProfiling::SendCollectionSlot() {
+  current_collected_slot_ = GetCurrentCollectionSlot();
+  if (current_collected_slot_ == last_checked_slot_) {
+    return;
+  }
+
+  MaybeResetEphemeralId();
+
+  auto resource_request = std::make_unique<network::ResourceRequest>();
+  resource_request->url = operational_profiling_endpoint_;
+  resource_request->headers.SetHeader("X-Brave-FL-Operational-Profile", "?1");
+
+  resource_request->credentials_mode = network::mojom::CredentialsMode::kOmit;
+  resource_request->method = "POST";
+
+  url_loader_ = network::SimpleURLLoader::Create(
+      std::move(resource_request), GetNetworkTrafficAnnotationTag());
+  url_loader_->AttachStringForUpload(BuildPayload(), "application/base64");
+
+  url_loader_->DownloadToStringOfUnboundedSizeUntilCrashAndDie(
+      url_loader_factory_.get(),
+      base::BindOnce(&BraveOperationalProfiling::OnUploadComplete,
+                     base::Unretained(this)));
+}
+
+void BraveOperationalProfiling::OnUploadComplete(
+    std::unique_ptr<std::string> response_body) {
+  if (url_loader_->ResponseInfo()->headers->response_code() == 200) {
+    last_checked_slot_ = current_collected_slot_;
+    SavePrefs();
+  }
+}
+
+std::string BraveOperationalProfiling::BuildPayload() const {
+  base::Value root(base::Value::Type::DICTIONARY);
+
+  root.SetKey("ephemeral_id", base::Value(ephemeral_id_));
+  root.SetKey("platform", base::Value(platform_));
+  root.SetKey("collection_slot", base::Value(current_collected_slot_));
+
+  std::string result;
+  base::JSONWriter::Write(root, &result);
+
+  return result;
+}
+
+int BraveOperationalProfiling::GetCurrentCollectionSlot() const {
+  base::Time::Exploded now;
+  base::Time::Now().LocalExplode(&now);
+
+  return ((now.day_of_month - 1) * 24 * 60 + now.hour * 60 + now.minute) /
+         collection_slot_size_in_minutes_;
+}
+
+std::string BraveOperationalProfiling::GetPlatformIdentifier() {
+  return brave_stats::GetPlatformIdentifier();
+}
+
+void BraveOperationalProfiling::MaybeResetEphemeralId() {
+  const base::Time now = base::Time::Now();
+  if (ephemeral_id_.empty() || (!ephemeral_id_expiration_time_.is_null() &&
+                                now > ephemeral_id_expiration_time_)) {
+    ephemeral_id_ =
+        base::ToUpperASCII(base::UnguessableToken::Create().ToString());
```

Comment:
Just a note that UnguessableToken is a 128-bit random value, not 256-bit as detailed in https://github.com/brave/security/issues/439#issuecomment-841463663 and the [spec](https://docs.google.com/document/d/15vqdMUMUQqdRUxPAXyC96i2ftcXb30YZ7VNkwM-XcoU/edit#heading=h.fpqk7mr7eryi). @orspetol I think this is fine but do you have any concerns?

Reply:
128-bits is more than enough for this purpose, however the spec needs to be updated to match this. We can use Base::RandBytes to generate a 256bit value if we need; however, IMO this would be overkill. Update the spec and we are good.

Reply:
Updated the spec 👍